rosbag2_transport/test/rosbag2_transport/test_record_topic_types.cpp

```diff
@@ -0,0 +1,124 @@
+    std::move(writer_), storage_options_, record_options);
+  recorder->record();
+
+  start_async_spin(recorder);
+
+  ASSERT_TRUE(pub_manager.wait_for_matched(array_topic_name.c_str()));
+  ASSERT_FALSE(pub_manager.wait_for_matched(string_topic_name.c_str()));
+
+  pub_manager.run_publishers();
+
+  auto & writer = recorder->get_writer_handle();
+  auto & mock_writer = dynamic_cast<MockSequentialWriter &>(writer.get_implementation_handle());
+
+  size_t expected_messages = 2;
+  auto ret = rosbag2_test_common::wait_until_shutdown(
+    std::chrono::seconds(5),
+    [&mock_writer, &expected_messages]() {
+      return mock_writer.get_messages().size() >= expected_messages;
+    });
+  EXPECT_TRUE(ret) << "failed to capture expected messages in time";
+  auto recorded_messages = mock_writer.get_messages();
+  EXPECT_EQ(recorded_messages.size(), expected_messages);
+  auto recorded_topics = mock_writer.get_topics();
+  EXPECT_THAT(recorded_topics, SizeIs(1));
+  EXPECT_TRUE(recorded_topics.find(array_topic_name) != recorded_topics.end());
+  EXPECT_TRUE(recorded_topics.find(string_topic_name) == recorded_topics.end());
+}
+
+
+TEST_F(RecordIntegrationTestFixture, all_overrides_topic_types)
+{
+  auto array_message = get_messages_arrays()[0];
+  array_message->float32_values = {{40.0f, 2.0f, 0.0f}};
+  array_message->bool_values = {{true, false, true}};
+  std::string array_topic_name = "/array_topic";
+
+  auto string_message = get_messages_strings()[0];
+  string_message->string_value = "Hello World";
+  std::string string_topic_name = "/string_topic";
+
+  rosbag2_test_common::PublicationManager pub_manager;
+  pub_manager.setup_publisher(array_topic_name, array_message, 2);
+  pub_manager.setup_publisher(string_topic_name, string_message, 2);
+
+  rosbag2_transport::RecordOptions record_options =
+  {true, false, false, {}, {"test_msgs/msg/Arrays"}, {}, {}, {}, "rmw_format", 100ms};
+
+  auto recorder = std::make_shared<rosbag2_transport::Recorder>(
+    std::move(writer_), storage_options_, record_options);
+  recorder->record();
+
+  start_async_spin(recorder);
+
+  ASSERT_TRUE(pub_manager.wait_for_matched(array_topic_name.c_str()));
+  ASSERT_TRUE(pub_manager.wait_for_matched(string_topic_name.c_str()));
+
+  pub_manager.run_publishers();
+
+  auto & writer = recorder->get_writer_handle();
+  auto & mock_writer = dynamic_cast<MockSequentialWriter &>(writer.get_implementation_handle());
+
+  size_t expected_messages = 2;
+  auto ret = rosbag2_test_common::wait_until_shutdown(
+    std::chrono::seconds(5),
+    [&mock_writer, &expected_messages]() {
+      return mock_writer.get_messages().size() >= expected_messages;
+    });
+  EXPECT_TRUE(ret) << "failed to capture expected messages in time";
+  auto recorded_messages = mock_writer.get_messages();
+  EXPECT_NE(recorded_messages.size(), expected_messages);
+  auto recorded_topics = mock_writer.get_topics();
+  EXPECT_THAT(recorded_topics, SizeIs(4));
```

Comment:
Why expect to get 4 topics? I see that we are publishing only on two topics `"/array_topic"` and `"/string_topic"`. 
```suggestion
  EXPECT_THAT(recorded_topics, SizeIs(2));
```
BTW. Curious how this test passes on the CI?